src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,146 @@
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/later.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace v8_engine {
+
+namespace internal {
+
+// spsc_queue
+
+spsc_queue::spsc_queue()
+  : _items(ss::smp::count)
+  , _ss_mutex_shard(ss::this_shard_id()) {}
+
+void spsc_queue::close() {
+    is_stopped = true;
+    _push_mutex.broken(ss::gate_closed_exception());
+    _has_element_cv.notify_all();
+}
+
+ss::future<> spsc_queue::push(work_item* item) {
+    auto lock = co_await _push_mutex.get_units();
+
+    while (!_items.push(item)) {
+        co_await _is_not_full.wait();
+    }
+
+    _has_element_cv.notify_one();
+    co_return;
+}
+
+work_item* spsc_queue::pop() {
+    work_item* item = nullptr;
+    std::unique_lock lock{_std_mutex};
+
+    // We need to use wait_for, because std::thread can lose exception from
+    // seastar thread (between check pred and sleep)
+    _has_element_cv.wait_for(lock, _timeout_cond_wait_std_thread_ms, [this] {
+        return !_items.empty() || is_stopped;
+    });
+
+    if (!_items.empty()) {
+        bool need_notify = _items.write_available() == 0;
+        item = _items.front();
+        _items.pop();
+
+        if (need_notify) {
+            _is_not_full.write_side().signal(1);
+        }
+    }
+
+    return item;
+}
+
+bool spsc_queue::empty() { return _items.empty(); }
+
+} // namespace internal
+
+// executor
+
+executor::executor(uint8_t cpu_id)
+  : _threads([this, cpu_id] {
+      pin(cpu_id);
+      loop();
+  })
+  , _watchdog_shard(ss::this_shard_id()) {}
+
+executor::~executor() { _threads.join(); }
+
+ss::future<> executor::stop() {
+    _tasks.close();
+    return _gate.close().then([this] { _is_stopped = true; });
+}
+
+bool executor::is_stopping() const {
+    return _is_stopped.load(std::memory_order_relaxed);
+}
+
+void executor::rearm_watchdog(
+  internal::work_item& task, std::chrono::milliseconds timeout) {
+    _watchdog.set_callback([&task] { task.cancel(); });
+
+    _watchdog.rearm(
+      ss::lowres_clock::time_point(ss::lowres_clock::now() + timeout));
+}
+
+void executor::cancel_watchdog(internal::work_item& task) {
+    if (!_watchdog.cancel()) {
+        task.on_timeout();
+    }
+}
+
+void executor::pin(unsigned cpu_id) {
+    cpu_set_t cs;
+    CPU_ZERO(&cs);
+    CPU_SET(cpu_id, &cs);
+    auto r = pthread_setaffinity_np(pthread_self(), sizeof(cs), &cs);
+    vassert(r == 0, "Can not pin executor thread to core {}", cpu_id);
+}
+
+void executor::loop() {
+    while (!(is_stopping() && _tasks.empty())) {
+        internal::work_item* item = _tasks.pop();
+
+        if (item) {
+            ss::alien::submit_to(_watchdog_shard, [this, item] {
+                rearm_watchdog(*item, item->get_timeout());
+                return ss::now();
```

Comment:
do we need to return `future<>` from this lambda ?

Reply:
Yes

```
template<typename Func, typename T = internal::return_type_t<Func>>
std::future<T> submit_to(unsigned shard, Func func) {
    std::promise<T> pr;
    auto fut = pr.get_future();
    run_on(shard, [pr = std::move(pr), func = std::move(func)] () mutable {
        // std::future returned via std::promise above.
        (void)func().then_wrapped([pr = std::move(pr)] (auto&& result) mutable {
            try {
                internal::return_type_of<Func>::set(pr, result.get());
            } catch (...) {
                pr.set_exception(std::current_exception());
            }
        });
    });
    return fut;
}
```

We need to return future from function in submit _to